device/simulation/tt_simulation_host.cpp

```diff
@@ -0,0 +1,79 @@
+// SPDX-FileCopyrightText: (c) 2023 Tenstorrent Inc.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#include <typeinfo>
+#include <sstream>
+#include <iomanip>
+#include <filesystem>
+#include <cassert>
+#include <cstdlib>
+
+#include "common/logger.hpp"
+#include "common/assert.hpp"
+#include "tt_simulation_host.hpp"
+
+tt_SimulationHost::tt_SimulationHost() {
+    // Get current date time string
+    std::time_t time = std::time(nullptr);
+    std::tm local_time = *std::localtime(&time);
+    char buffer[100];
+    std::strftime(buffer, sizeof(buffer), "%m-%d-%H:%M:%S", &local_time);
+
+    // Get socket name
+    const char *nng_socket_name = std::getenv("NNG_SOCKET_NAME") ? std::getenv("NNG_SOCKET_NAME") : "nng_ipc";
+    const char *user_name = std::getenv("USER");
+
+    std::ostringstream ss;
+    ss << NNG_SOCKET_PREFIX << user_name << "_" << buffer << "_" << nng_socket_name;
+    std::string nng_socket_addr_str = ss.str();
+    const char *nng_socket_addr = nng_socket_addr_str.c_str();
+    setenv("NNG_SOCKET_ADDR", nng_socket_addr, 1);  // pass NNG_SOCKET_ADDR to remote
+
+    // Open socket and create dialer
+    log_info(tt::LogEmulationDriver, "Dialing: {}", nng_socket_addr);
+    nng_pair1_open(&host_socket);
+    int rv = nng_dialer_create(&host_dialer, host_socket, nng_socket_addr);
+    TT_ASSERT(rv == 0, "Failed to create dialer: {} {}", nng_strerror(rv), nng_socket_addr);
+}
+
+tt_SimulationHost::~tt_SimulationHost() {
+    nng_dialer_close(host_dialer);
+    nng_close(host_socket);
+}
+
+void tt_SimulationHost::start_host() {
+    // Establish connection with remote VCS simulator
+    int rv;
+    do {
+        rv = nng_dialer_start(host_dialer, 0);
+        if (rv != 0) {
+            log_info(tt::LogEmulationDriver, "Waiting for remote: {}", nng_strerror(rv));
+            std::this_thread::sleep_for(std::chrono::seconds(1));
+        }
+    } while (rv != 0);
+}
+
+void tt_SimulationHost::send_to_device(uint8_t *buf, size_t buf_size) {
+    int rv;
+    log_debug(tt::LogEmulationDriver, "Sending messsage to remote..");
+    
+    void *msg = nng_alloc(buf_size);
+    std::memcpy(msg, buf, buf_size);
+
+    rv = nng_send(host_socket, msg, buf_size, NNG_FLAG_ALLOC);
+    log_debug(tt::LogEmulationDriver, "Message sent.");
+    if (rv != 0) {
+        log_info(tt::LogEmulationDriver, "Failed to send message to remote: {}", nng_strerror(rv));
+    }
+}
+
+void tt_SimulationHost::recv_from_device(void **data_ptr, size_t *data_size) {
```

Comment:
why is `data_size` a pointer? this is odd

Reply:
Have you considered changing the function signature to:

`size_t tt_SimulationHost::recv_from_device(void **data_ptr)`

Reply:
Passing a ptr since the nng api also takes a pointer `int nng_recv(nng_socket s, void *data, size_t *sizep, int flags);`

Should I just pass no ptr?

Reply:
Changed to `size_t tt_SimulationHost::recv_from_device(void **data_ptr)`